docs/topics/addons.rst

```diff
@@ -0,0 +1,190 @@
+            settings.set("DNSCACHE_ENABLED", True, "addon")
+
+This allows users to override these settings in the project or spider
+configuration. This is not possible with settings that are mutable objects,
+such as the dict that is a value of :setting:`ITEM_PIPELINES`. In these cases
+you can provide an add-on-specific setting that governs whether the add-on will
+modify :setting:`ITEM_PIPELINES`::
+
+    class MyAddon:
+        def update_settings(self, settings):
+            if settings.getbool("MYADDON_ENABLE_PIPELINE"):
+                settings["ITEM_PIPELINES"]["path.to.mypipeline"] = 200
+
+Fallbacks
+---------
+
+Some components provided by add-ons need to fall back to "default"
+implementations, e.g. a custom download handler needs to send the request that
+it doesn't handle via the default download handler, or a stats collector that
+includes some additional processing but otherwise uses the default stats
+collector. And it's possible that a project needs to use several custom
+components of the same type, e.g. two custom download handlers that support
+different kinds of custom requests and still need to use the default download
+handler for other requests. To make such use cases easier to configure, we
+recommend that such custom components should be written in the following way:
+
+1. The custom component (e.g. ``MyDownloadHandler``) shouldn't inherit from the
+   default Scrapy one (e.g.
+   ``scrapy.core.downloader.handlers.http.HTTPDownloadHandler``), but instead
+   be able to load the class of the fallback component from a special setting
+   (e.g. ``MY_FALLBACK_DOWNLOAD_HANDLER``), create an instance of it and use
+   it.
+2. The add-ons that include these components should read the current value of
+   the default setting (e.g. ``DOWNLOAD_HANDLERS``) in their
+   ``update_settings()`` methods, save that value into the fallback setting
+   (``MY_FALLBACK_DOWNLOAD_HANDLER`` mentioned earlier) and set the default
+   setting to the component provided by the add-on (e.g.
+   ``MyDownloadHandler``). If the fallback setting is already set by the user,
+   they shouldn't change it.
+3. This way, if there are several add-ons that want to modify the same setting,
+   all of them will fallback to the component from the previous one and then to
+   the Scrapy default. The order of that depends on the priority order in the
+   ``ADDONS`` setting.
+
+
+Add-on examples
+===============
+
+Set some basic configuration:
+
+.. code-block:: python
+
+    class MyAddon:
+        def update_settings(self, settings):
+            settings["ITEM_PIPELINES"]["path.to.mypipeline"] = 200
+            settings.set("DNSCACHE_ENABLED", True, "addon")
+
+Check dependencies:
+
+.. code-block:: python
+
+    class MyAddon:
+        def update_settings(self, settings):
+            try:
+                import boto
+            except ImportError:
+                raise RuntimeError("MyAddon requires the boto library")
```

Comment:
What would happen if NotConfigured error is raised? I think this is the error which components raise when a dependency is not available. Is there a reason to prefer RuntimeError in addons?

Reply:
I have no strong opinion on it, but I think it makes sense that, if you have enabled a plugin that requires a dependency, and that dependency is missing, your crawler should fail to initialize, forcing you to either disable the plugin or install the missing dependency. Same for missing required settings.

I see `NotConfigured` as something a Scrapy component raises when it is configured but disabled through settings.

If a plugin has an `*_ENABLED` setting, then I think it would make sense to support that plugin raising `NotConfigured`, and that causing the plugin to not be used at run time, and a log message indicating so during initialization, but things working otherwise.

So I think it may make sense to add support for `NotConfigured` on the addon manager class, but I think in some cases raising something else makes sense, including this one.

Reply:
So, the difference is that RuntimeError stops the spider, but NotConfigured doesn't? Does it work for addons the same way?

I'm not really familiar with these parts of the code, but we're using NotConfigured in similar cases:

* https://github.com/scrapy/scrapy/blob/7522aeed357d90ed95ee10d3b5a506f1b1049d1f/scrapy/pipelines/images.py#L61
* https://github.com/scrapy/scrapy/blob/7522aeed357d90ed95ee10d3b5a506f1b1049d1f/scrapy/extensions/feedexport.py#L176
* https://github.com/scrapy/scrapy/blob/7522aeed357d90ed95ee10d3b5a506f1b1049d1f/scrapy/core/downloader/handlers/s3.py#L20

Reply:
I think that’s because those components are configured by default, i.e. the `_ENABLED=True` is implicit, so doing otherwise would be a bad user experience. But for opt-in components, including add-ons, I think stopping the spider may be best.

Reply:
OK, so nothing handles `NotConfigured` raised there.  Which is expected because it's only handled in `MiddlewareManager` (which is the base for middleware, pipeline and extension managers) and `DownloadHandlers`. We should probably add handling to AddonManager to similarly silently skip add-ons that raise it?

Reply:
If we do that I wonder if we should merge `AddonManager.update_settings()` into `AddonManager.load_settings()` or into `AddonManager._add()` so that `AddonManager.addons` never contains add-ons that raise `NotConfigured`. Also it doesn't seem useful to have a separate function that calls add-ons' `update_settings()` as we don't have any features that would benefit from it.